File: 2022/Day16ByPart/CaveLayout.cs
internal class CaveLayout
{

    public Dictionary<string, Node> Nodes = new Dictionary<string, Node>();

    public Dictionary<string, int> Memo = new Dictionary<string, int>();
    public long AllOpen = 0;

    public CaveLayout(List<string> inputCol)
    {
        List<Tuple<string, string>> listOfInputTup = new List<Tuple<string, string>>();

        string strTmp;
        string[] strTmpArr;
        string name;
        int flow;
        string nodes;
        int index = 0;
        foreach (string str in inputCol)
        {
            strTmp = str.Replace("Valve ", "").Trim();
            strTmpArr = strTmp.Split(" has flow rate=");
            name = strTmpArr[0];
            if (strTmpArr[1].Contains("tunnels lead to valves"))
                strTmpArr = strTmpArr[1].Split("; tunnels lead to valves ");
            else
                strTmpArr = strTmpArr[1].Split("; tunnel leads to valve ");

            flow = int.Parse(strTmpArr[0]);
            nodes = strTmpArr[1];

            listOfInputTup.Add(new Tuple<string, string>(name, nodes));
            Nodes.Add(name, new Node(name, flow, index));
            if (flow > 0)
                AllOpen |= (1 << index);
            index++;
        }
        string[] nodesArr;

        foreach (Tuple<string, string> tup in listOfInputTup)
        {
            name = tup.Item1;
            nodesArr = tup.Item2.Split(",");
            Node baseNode = Nodes[name];
            foreach (string str in nodesArr)
                baseNode.AddAdjenced(Nodes[str.Trim()]);
        }

    }


    internal void ResetNodes()
    {
        foreach (Node node in Nodes.Values)
        {
            node.Distance = -1;
            node.Attended = false;
        }
    }
}

File: 2022/Day16++/CaveLayout.cpp
#pragma once
using namespace std;
#include"CaveLayout.h"
#include"Node.h"

CaveLayout::CaveLayout(vector<string>& inputCol) {


	vector<pair<string, string>> pairs;

	for (string str : inputCol) {
		
		std::shared_ptr<Node> nPtr = std::shared_ptr<Node>(new Node(str, pairs));
        Nodes[(nPtr->Name)] = nPtr;
	}

	for (pair<string, string>  const& par : pairs)
	{
		Nodes[par.first]->AddAdjenced(Nodes[par.second]);
	}
}

void CaveLayout::ResetNodes() {
	for(pair<string, std::shared_ptr<Node>>  const& pair : Nodes)
	{
		pair.second->Distance = -1;
		pair.second->Attended = false;
	}
}